File: sharedUI/src/commonMain/kotlin/org/alphakids/app/onboarding/LoginViewModel.kt
package org.alphakids.app.onboarding

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import org.alphakids.app.onboarding.domain.model.LoginRequest
import org.alphakids.app.onboarding.domain.repository.AuthRepository
import org.alphakids.app.parent.domain.repository.ParentRepository

/**
 * UI state for the login screen.
 */
data class LoginUiState(
    val email: String = "",
    val password: String = "",
    val isLoading: Boolean = false,
    val error: String? = null,
    val isSuccess: Boolean = false,
    val hasChildren: Boolean = false,
)

/**
 * ViewModel for [LoginScreen].
 *
 * Handles email/password input, validation, and authentication through [AuthRepository].
 */
class LoginViewModel(
    private val authRepository: AuthRepository,
    private val parentRepository: ParentRepository,
) : ViewModel() {

    private val _uiState = MutableStateFlow(LoginUiState())
    val uiState: StateFlow<LoginUiState> = _uiState.asStateFlow()

    fun onEmailChanged(value: String) {
        _uiState.update { it.copy(email = value, error = null) }
    }

    fun onPasswordChanged(value: String) {
        _uiState.update { it.copy(password = value, error = null) }
    }

    fun onLoginClick() {
        val state = _uiState.value

        // Validate email
        if (state.email.isBlank()) {
            _uiState.update { it.copy(error = "El email es requerido") }
            return
        }
        if (!isValidEmail(state.email)) {
            _uiState.update { it.copy(error = "Formato de email inválido") }
            return
        }

        // Validate password
        if (state.password.isBlank()) {
            _uiState.update { it.copy(error = "La contraseña es requerida") }
            return
        }

        _uiState.update { it.copy(isLoading = true, error = null, isSuccess = false) }

        viewModelScope.launch {
            val response = authRepository.login(
                LoginRequest(email = state.email, password = state.password),
            )

            if (response.success) {
                val children = parentRepository.getChildren()
                _uiState.update {
                    it.copy(isLoading = false, isSuccess = true, hasChildren = children.isNotEmpty())
                }
            } else {
                _uiState.update { it.copy(isLoading = false, error = response.message) }
            }
        }
    }

    fun onRegisterClick() {
        _uiState.update { it.copy(isSuccess = true) }
    }

    private fun isValidEmail(email: String): Boolean {
        return email.matches(Regex("^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$"))
    }
}

File: iosApp/iosApp/ViewModels/LoginViewModel.swift
import Foundation
import SharedLogic

// MARK: - LoginViewModel

/// ViewModel for the iOS LoginScreen.
///
/// Mirrors the business logic from the shared Kotlin `LoginViewModel` at:
/// `sharedUI/src/commonMain/kotlin/org/alphakids/app/onboarding/LoginViewModel.kt`
///
/// Key behaviors:
/// - Email validation: not blank, contains `@` and `.`, min 5 chars
/// - Password validation: not blank
/// - On successful login (`response.success == true`), sets `isLoggedIn = true`
/// - On failure, sets `errorMessage` to `response.message`
@MainActor
final class LoginViewModel: ObservableObject {

    // MARK: - Published State

    @Published var email: String = ""
    @Published var password: String = ""
    @Published var isLoading: Bool = false
    @Published var errorMessage: String? = nil
    @Published var isLoggedIn: Bool = false

    // MARK: - Computed Properties

    /// Returns `true` when email and password pass basic validation.
    /// Allows the "Iniciar sesión" button to be enabled.
    var isFormValid: Bool {
        let trimmedEmail = email.trimmingCharacters(in: .whitespaces)
        return !trimmedEmail.isEmpty
            && isValidEmail(trimmedEmail)
            && !password.isEmpty
    }

    // MARK: - Public API

    /// Triggers the login flow.
    ///
    /// 1. Validates email and password (same rules as Kotlin `LoginViewModel`).
    /// 2. Calls `AuthRepositoryAsync.login(email:password:)`.
    /// 3. On success → sets `isLoggedIn = true`.
    /// 4. On failure → sets `errorMessage` to the server response message.
    func login() async {
        let trimmedEmail = email.trimmingCharacters(in: .whitespaces)

        // Validate email
        guard !trimmedEmail.isEmpty else {
            errorMessage = "El email es requerido"
            return
        }
        guard isValidEmail(trimmedEmail) else {
            errorMessage = "Formato de email inválido"
            return
        }

        // Validate password
        guard !password.isEmpty else {
            errorMessage = "La contraseña es requerida"
            return
        }

        // Clear previous errors, show loading
        isLoading = true
        errorMessage = nil

        do {
            let response = try await AuthRepositoryAsync.login(
                email: trimmedEmail,
                password: password
            )

            // response.success mirrors AuthResponse.success from the Kotlin model
            if response.success {
                isLoading = false
                isLoggedIn = true
            } else {
                isLoading = false
                errorMessage = response.message
            }
        } catch {
            isLoading = false
            errorMessage = error.localizedDescription
        }
    }

    // MARK: - Private Helpers

    /// Email validation matching the Kotlin `LoginViewModel.isValidEmail()`.
    ///
    /// Rules: must contain `@` AND `.`, and be at least 5 characters.
    private func isValidEmail(_ email: String) -> Bool {
        let cleaned = email.trimmingCharacters(in: .whitespaces).lowercased()
        return cleaned.contains("@") && cleaned.contains(".") && cleaned.count >= 5
    }
}
